### ISSUE TITLE
FunctionalPreprocessingStage.adapt doesn't seem to actually adapt.

### ISSUE THREAD

**Issue by tibbe** [keras-team/keras#15219]
------------------------

(Originally filed as https://github.com/tensorflow/tensorflow/issues/50820 with more context.)

### System information

Google Colab

-   **Have I written custom code (as opposed to using a stock example script
    provided in Keras)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
-   **TensorFlow installed from (source or binary)**: Binary
-   **TensorFlow version (use command below)**: 2.6.0
-   **Python version**: 3.7.11
-   **Bazel version (if compiling from source)**:
-   **GPU model and memory**: Google colab
-   **Exact command to reproduce**:

### Describe the problem

While the `FunctionalPreprocessingStage.adapt()` method runs without errors it doesn't seem to work. 

### Source code / logs

If I individually `adapt` layers like so:

```py
inputs = {}
normalization_layers = {}

for name in NUMERIC_FEATURES:
  inputs[name] = layers.Input(shape=(1,))
  normalization_layers[name] = preprocessing.Normalization()

for name, layer in normalization_layers.items():
  layer.adapt(train_dataset.map(lambda x, y: x[name]))

input_layers = []
for name, input in inputs.items():
  input_layers.append(normalization_layers.get(name)(input))
x = tf.keras.layers.Concatenate()(input_layers)
x = layers.Dense(32)(x)
x = layers.Dense(16)(x)
output = layers.Dense(1)(x)
model = keras.Model(inputs=inputs, outputs=output)
model.compile(optimizer='adam', loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError(name='rmse')])
```

the model trains (the loss is decreasing). If I instead use the `FunctionalPreprocessingState` helper class like so:

```py
inputs = []
normalization_layers = []

for name in NUMERIC_FEATURES:
  input = layers.Input(shape=(1,))
  inputs.append(input)
  normalization_layers.append(preprocessing.Normalization()(input))

preproc_layer = preprocessing_stage.FunctionalPreprocessingStage(
    inputs, normalization_layers)
preproc_layer.adapt(train_dataset)

x = tf.keras.layers.Concatenate()(preproc_layer.outputs)
x = layers.Dense(32, activation="relu")(x)
x = layers.Dense(16, activation="relu")(x)
output = layers.Dense(1)(x)
model = keras.Model(inputs=preproc_layer.inputs, outputs=output)
model.compile(optimizer='adam', loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError(name='rmse')])
```

It doesn't train (the loss isn't decreasing). I suspect that the layers didn't get adapted properly and are doing random transformations to the input features during training.

**Comment by mattdangerw** [comment]
Thanks for the bug! We will take a look. Worth noting that the `FunctionalPreprocessingStage` is not an exported symbol of Keras, experimental or otherwise. For now, the first code sample you posted (calling adapt individually on each layer) is the correct approach.

We are definitely aware that having no convenient way to adapt a set of layers at once is a big usability problem, and are looking at it. But we are still exploring exactly how we want to expose the functionality to adapt a set of layers (e.g. a PreprocessingStage class, something directly on keras.Model, something else). That class is still very much a work in progress.

**Comment by mattdangerw** [comment]
So the latest on this, we do still want to support a good API for adapting multiple layers, but the current thinking is to do that directly on the model class, and not with a separate FunctionalPreprocessingStage. Closing this, as FunctionalPreprocessingStage is not an exported keras symbol.
